### PR TITLE
feat: Compress uploads

### LINKED ISSUE #285: Minidumps are not compressed before upload
- [x] Review the documentation: https://docs.sentry.io/platforms/javascript/electron/
- [x] Search for existing issues: https://github.com/getsentry/sentry-electron/issues
- [x] Use the latest release: https://github.com/getsentry/sentry-electron/releases
- [ ] Provide a link to the affected event from your Sentry account

## Versions + Platform

- [x] SDK version - `@sentry/electron@v2.0.4`
- [x] Electron version - `electron@v8.2.4`
- [x] Platform - `Windows 10`

## Description

Documentation mentions 20MB limit for a compressed minidump request: https://docs.sentry.io/platforms/native/guides/minidumps/#size-limits

But it seems my minidump request is not beeing compressed. Sentry successfully uploads ~19MB minidumps on app restart (after crash), but fails to upload 24 MB minidump **(Sentry Logger [Warn]: Failed to upload minidump SentryError: HTTP Error (413))**. Why are you not compressing minidumps? Documentation says you support compressed minidump uploads. After compression, my 24MB minidump becomes a mere 240 KB zip file.

### PR DESCRIPTION
**Pull request by timfish**
Fixes #285

### CODE REVIEW

File: src/main/transports/net.ts [hunk 1]
```diff
@@ -117,8 +119,11 @@ export class NetTransport extends Transports.BaseTransport {
             // Drain
           });
         });
-        req.write(request.body);
-        req.end();
+
+        gzip(request.body, (_, compressedBody) => {
+          req.write(compressedBody);
```
Comment (jan-auer) [inline src/main/transports/net.ts:124]: This is fine. If you like to reduce overheads here, you could also create a gzip stream and then have it flush into the request, like described [here](https://nodejs.org/api/zlib.html#zlib_flushing).

**Review by jan-auer — APPROVED** [review]
Thank you, @timfish! LGTM. 

As a suggestion, you could only trigger compression if the body exceeds a certain size, since gzip also comes with a CPU penalty.

**Comment by timfish** [comment]
> if the body exceeds a certain size

What would be a good threshold?

**Comment by jan-auer** [comment]
I did some quick research, and without explicit benchmarking the threshold is hard to determine. Most resources I found talk about a limit lower than 1k, which would be almost every outgoing request.

There is also https://medium.com/@joehonton/the-gzip-penalty-d31bd697f1a2, which seems to tip over at 32k. This gets me thinking: We probably don't need compression for regular event submission, and events are usually under 32k, so let's go with that?

**Comment by HazAT** [comment]
So `if(request.body.length > 32000) zip else not zip`?

**Comment by timfish** [comment]
Now with streams and gzip threshold!

File: src/main/transports/net.ts [hunk 2]
```diff
@@ -2,17 +2,34 @@ import { Transports } from '@sentry/node';
 import { Event, Response, SentryRequest, Status, TransportOptions } from '@sentry/types';
 import { parseRetryAfterHeader, PromiseBuffer, SentryError, timestampWithMs } from '@sentry/utils';
 import { net } from 'electron';
+import { Readable, Writable } from 'stream';
 import * as url from 'url';
+import { createGzip } from 'zlib';
 
 import { isAppReady } from '../backend';
 
+const GZIP_THRESHOLD = 1024 * 32;
```
Comment (jan-auer) [inline src/main/transports/net.ts:11]: _nit_: I'd mention in a comment that this is an estimated maximum size for reasonable standalone events.

**Review by jan-auer — APPROVED** [review]
Looks great, thanks for going the extra mile.

**Comment by GintV** [comment]
Any particular reason why this one is not merged yet?

**Comment by GintV** [comment]
Thanks for implementing this @timfish! @jan-auer, @HazAT any chance release could be made with this change?

**Pull request merged**